### DISCUSSION TITLE
Callable type mismatch

### DISCUSSION THREAD

**Discussion by ImogenBits** [microsoft/pyright#5569] [Q&A] [answered]
This code doesn't type check:
```py
class FnType(Protocol):
    def __call__(self, a: int) -> int: ...

def f(b: int = 0, a: int = 0) -> int: ...

var: FnType = f
```
I'm not sure if this is a bug or intended behaviour? I can see how `f` appears to not have the correct type since the first parameter has a different name, but why does that matter if you can still call `f(a=0)` since the other parameter has that name? Is this just a limitation of the python type system? This feels really strange since if you change the `var` annotation to `Callable[[int], int]` the code gets accepted, so this kind of callable subtyping seems natural.

**Comment by Kaniee** [comment, marked as answer]
It's because `var(3)` would call `f` with `b=3` and `a=0`.

You can change the Protocol such that `a` has to be passed as a keyword argument:
```
class FnType(Protocol):
    def __call__(self, *, a: int) -> int:
        ...
```

With this `var(3)` would not be allowed and the function call would be narrowed down to `var(a=3)`.
